### PR TITLE
feat(types,utils): add promotions create with application method

### PR DESCRIPTION
**Pull request by riqwan**
What:

- Promotions can be created with its bare attributes
- Promotions one to one relationship with ApplicationMethod can be created with its attributes + validation

RESOLVES CORE-1592
RESOLVES CORE-1595

### CODE REVIEW

File: packages/promotion/src/migrations/Migration20231220153431.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { Migration } from "@mikro-orm/migrations"
+
+export class Migration20231220153431 extends Migration {
+  async up(): Promise<void> {
+    this.addSql(
+      'create table "promotion" ("id" text not null, "code" text not null, "is_automatic" boolean not null default false, "type" text check ("type" in (\'standard\', \'buyget\')) not null, "created_at" timestamptz not null default now(), "updated_at" timestamptz not null default now(), "deleted_at" timestamptz null, constraint "promotion_pkey" primary key ("id"));'
```
Comment (adrien2p) [inline packages/promotion/src/migrations/Migration20231220153431.ts]: suggestion: We should probably add an index on code as well

File: packages/promotion/src/models/application-method.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  })
+  updated_at?: Date
+
+  @Property({ columnType: "timestamptz", nullable: true })
+  deleted_at?: Date
+
+  @BeforeCreate()
+  onCreate() {
+    this.id = generateEntityId(this.id, "app_method")
+  }
```
Comment (adrien2p) [inline packages/promotion/src/models/application-method.ts:75]: todo: add onInit

File: packages/promotion/src/models/promotion.ts [hunk 3]
```diff
@@ -1,16 +1,71 @@
...
 export default class Promotion {
+  [OptionalProps]?: OptionalFields | OptionalRelations
+
   @PrimaryKey({ columnType: "text" })
   id!: string
 
+  @Property({ columnType: "text" })
+  @Unique({
+    name: "IDX_promotion_code_unique",
+    properties: ["code"],
```
Comment (adrien2p) [inline packages/promotion/src/models/promotion.ts:30]: suggestion: add index as well, constraint and index does not act exactly the same

File: packages/types/src/promotion/common/application-method.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  type: ApplicationMethodType
+  target_type: ApplicationMethodTargetType
+  allocation?: ApplicationMethodAllocation
+  value?: number
+  max_quantity?: number
+  promotion?: PromotionDTO | string
+}
+
+export interface UpdateApplicationMethodDTO {
+  id: string
```
Comment (adrien2p) [inline packages/types/src/promotion/common/application-method.ts:22]: question: we will add the necessary properties later?

Comment (adrien2p) [reply packages/types/src/promotion/common/application-method.ts:22]: same question for all update interfaces that only refer to id

**Review by adrien2p — APPROVED** [review]
LGTM, few comments only and one todo

File: packages/types/src/promotion/common/application-method.ts (hunk 4 above)
Comment (riqwan) [reply packages/types/src/promotion/common/application-method.ts:22]: yes, this is just a placeholder for now, will fill them up when I get to update tickets

File: packages/promotion/src/models/application-method.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+  @OnInit()
+  onInit() {
+    this.id = generateEntityId(this.id, "promo")
```
Comment (olivermrbl) [inline packages/promotion/src/models/application-method.ts:79]: question: Shouldn't the prefix be `app_method`?

Comment (riqwan) [reply packages/promotion/src/models/application-method.ts:79]: ah damn, good catch. I've changed it in my current branch.

**Pull request merged**